File: src/main/java/io/schell/llm/benchmark/agent/ClaudeMessageProcessor.java
package io.schell.llm.benchmark.agent;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.function.Consumer;

public class ClaudeMessageProcessor implements Consumer<String> {
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Consumer<String> outputConsumer;

    public ClaudeMessageProcessor(Consumer<String> outputConsumer) {
        this.outputConsumer = outputConsumer;
    }

    @Override
    public void accept(String s) {
        try {
            JsonNode jsonNode = objectMapper.readTree(s);
            if (jsonNode.has("type")) {
                String type = jsonNode.get("type").asText();
                switch (type) {
                    case "stream_event" -> {
                        JsonNode event = jsonNode.get("event");
                        if (event.has("type")) {
                            switch (event.get("type").asText()) {
                                case "message_start" -> {
                                    JsonNode message = event.get("message");
                                    JsonNode content = message.get("content");
                                    if (content.isArray()) {
                                        for (JsonNode item : content) {
                                            String contentType = item.get("type").asText();
                                            if (contentType.equals("thinking")) {
                                                print(item.get("thinking").asText());
                                            } else if (contentType.equals("text")) {
                                                print(item.get("text").asText());
                                            } else {
                                                print(s);
                                            }
                                        }
                                        return;
                                    } else {
                                        String contentType = content.get("type").asText();
                                        if (contentType.equals("thinking")) {
                                            print(content.get("thinking").asText());
                                            return;
                                        } else if (contentType.equals("message")) {
                                            // print(content.get("message"));
                                            // JsonNode messageContent = content.get("content");
                                            // for now ignore these, they seem to be empty most of the time
                                            return;
                                        } else {
                                            println(s);
                                            return;
                                        }
                                    }
                                }
                                case "message_delta" -> {
                                    JsonNode delta = event.get("delta");
                                    // stop_reason=tool_use?
                                    // println(delta.get("delta").asText());
                                    return;
                                }
                                case "content_block_delta" -> {
                                    // {"type":"stream_event","event":{"type":"content_block_delta","index":2,"delta":{"type":"input_json_delta","partial_json":"\":"}},"session_id":"a03a9304-8ecb-4204-9c24-2201e8dda43e","parent_tool_use_id":null,"uuid":"b39c1d17-df4b-4a4a-877b-f7e0d754f45e"}
                                    JsonNode delta = event.get("delta");
                                    String deltaType = delta.get("type").asText();
                                    switch (deltaType) {
                                        case "thinking_delta" -> {
                                            print(delta.get("thinking").asText());
                                            return;
                                        }
                                        case "input_json_delta" -> {
                                            // print(delta.get("partial_json").asText());
                                            return;
                                        }
                                        case "signature_delta" -> {
                                            // we do not need to see this in the console log
                                            return;
                                        }
                                        case "text_delta" -> {
                                            print(delta.get("text").asText());
                                            return;
                                        }
                                        default -> {
                                            println(s);
                                            return;
                                        }
                                    }
                                }
                                case "content_block_start" -> {
                                    // ignore
                                    return;
                                }
                                case "content_block_stop" -> {
                                    // ignore
                                    return;
                                }
                                case "message_stop" -> {
                                    // ignore
                                    // TODO: this is probably where we want to put in the newlines
                                    println("");
                                    return;
                                }
                                default -> {
                                    print(s);
                                    return;
                                }
                            }
                        } else {
                            println(s);
                            return;
                        }
                    }
                    case "assistant" -> {
                        JsonNode message = jsonNode.get("message");
                        JsonNode content = message.get("content");
                        if (content.isArray()) {
                            for (JsonNode item : content) {
                                String itemType = item.get("type").asText();
                                switch (itemType) {
                                    case "thinking" -> {
                                        print(item.get("thinking").asText());
                                    }
                                    case "tool_use" -> {
                                        render_tool_use(item);
                                    }
                                    case "text" -> {
                                        print(item.get("text").asText());
                                    }
                                    default -> {
                                        print(s);
                                    }
                                }
                            }
                            return;
                        } else {
                            String itemType = content.get("type").asText();
                            switch (itemType) {
                                case "thinking" -> {
                                    print(content.get("thinking").asText());
                                    return;
                                }
                                case "tool_use" -> {
                                    render_tool_use(content);
                                    return;
                                }
                                case "text" -> {
                                    print(content.get("text").asText());
                                    return;
                                }
                                default -> {
                                    println(s);
                                    return;
                                }
                            }
                        }
                    }
                    case "user" -> {
                        JsonNode message = jsonNode.get("message");
                        JsonNode content = message.get("content");
                        if (content.isArray()) {
                            for (JsonNode item : content) {
                                String itemType = item.get("type").asText();
                                switch (itemType) {
                                    case "text" -> {
                                        println(item.get("text").asText());
                                        return;
                                    }
                                    case "tool_result" -> {
                                        if (item.has("content")) {
                                            JsonNode tool_result_content = item.get("content");
                                            if (tool_result_content.isTextual()) {
                                                String tool_result_string = tool_result_content.asText();
                                                String tool_result_with_newlines = tool_result_string.replaceAll("\\n", "\n");
                                                println("tool_result:\n" + tool_result_with_newlines);
                                                return;
                                            }
                                        }
                                        println(s);
                                        return;
                                    }
                                    default -> {
                                        println(s);
                                        return;
                                    }
                                }
                            }
                        } else {
                            String itemType = content.get("type").asText();
                            switch (itemType) {
                                case "text" -> {
                                    println(content.get("text").asText());
                                    return;
                                }
                                case "tool_result" -> {
                                    println("tool_result: " + content.get("content").toString());
                                    return;
                                }
                                default -> {
                                    println(s);
                                    return;
                                }
                            }
                        }
                        println(s);
                        return;
                    }
                    case "system" -> {
                        // ignore system message for now
                        println(s);
                        return;
                    }
                    case "result" -> {
//                            String subType = jsonNode.get("subtype").asText();
//                            boolean isError = jsonNode.get("is_error").asBoolean();
//                            long duration_ms = jsonNode.get("duration_ms").asLong();
//                            long duration_api_ms = jsonNode.get("duration_api_ms").asLong();
//                            int num_turns = jsonNode.get("num_turns").asInt();
//                            String result = jsonNode.get("result").asText();
//                            double total_cost_usd = jsonNode.get("total_cost_usd").asDouble();
//                            JsonNode usage = jsonNode.get("usage");
//                            long input_tokens = usage.get("input_tokens").asLong();
//                            long output_tokens = usage.get("output_tokens").asLong();
//                            JsonNode modelUsage = jsonNode.get("modelUsage");
//                            String permission_denials = jsonNode.get("permission_denials").asText();
                        println("Result: " + jsonNode.toPrettyString());
                    }
                    default -> {
                        println(s);
                        return;
                    }
                }
            }
        } catch (JsonProcessingException e) {
            // not a json log, probably a warning from claude, so we just output it
        }
        println(s);
    }

    private void render_tool_use(JsonNode item) {
        try {
            String name = item.get("name").asText();
            switch (name) {
                case "Edit" -> {
                    JsonNode input = item.get("input");
                    boolean replace_all = input.get("replace_all").asBoolean();
                    String file_path = input.get("file_path").asText();
                    String old_string = input.get("old_string").asText();
                    String new_string = input.get("new_string").asText();
                    println("Edit " + file_path);
                    println("Old");
                    println(old_string.replaceAll("\\n", "\n"));
                    println("New");
                    println(new_string.replaceAll("\\n", "\n"));
                }
                case "Glob" -> {
                    String pattern = item.get("input").get("pattern").asText();
                    println("\ntool_use: Glob " + pattern);
                }
                case "Read" -> {
                    String file_path = item.get("input").get("file_path").asText();
                    println("\ntool_use: Read " + file_path);
                }
                case "Write" -> {
                    String file_path = item.get("input").get("file_path").asText();
                    var input = item.get("input");

                    if (input.has("content")) {
                        var content = input.get("content").asText();
                        println("\ntool_use: Write " + file_path);
                        println("Content: ");
                        content = content.replaceAll("\\n", "\n");
                        println(content);
                    }

                }
                case "Bash" -> {
                    boolean run_in_background = false;
                    if (item.get("input").has("run_in_background")) {
                        run_in_background = item.get("input").get("run_in_background").asBoolean();
                    }
                    String description = "";
                    if (item.get("input").has("description")) {
                        description = item.get("input").get("description").asText();
                    }
                    String command = item.get("input").get("command").asText();
                    println("\ntool_use: Bash " + (run_in_background ? "(in background) " : " ") + description);
                    println("Command: " + command);
                }
                case "TaskOutput" -> {
                    String input = item.get("input").toString();
                    println("\ntool_use: " + name + " " + input);
                }
                case "TodoWrite" -> {
                    println("\ntool_use: TodoWrite");
                    JsonNode input = item.get("input");
                    JsonNode todos = input.get("todos");
                    for (JsonNode todo : todos) {
                        String content = todo.get("content").asText();
                        String status = todo.get("status").asText();
                        switch (status) {
                            case "in_progress" -> {
                                println(String.format("[⟳] %s", content));
                            }
                            case "pending" -> {
                                println(String.format("[⌛] %s", content));
                            }
                            case "completed" -> {
                                println(String.format("[✅] %s", content));
                            }
                            default -> {
                                println(String.format("[ ] %s", content));
                            }
                        }

                    }
                }
                default -> {
                    String input = item.get("input").toString();
                    println("\ntool_use: " + name + " " + input);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void print(String s) {
        System.out.print(s);
        if (outputConsumer != null) {
            outputConsumer.accept(s);
        }
    }

    private void println(String s) {
        System.out.println(s);
        if (outputConsumer != null) {
            outputConsumer.accept(s + "\n");
        }
    }
}

File: crates/benchmark-core/src/agent/claude_message_processor.rs
//! ClaudeMessageProcessor - parses Claude Code stream-json output
//! Logs key events with tracing and forwards formatted output to consumer for web UI streaming.

use serde_json::Value;
use tracing::{info, debug};

/// Callback type for sending processed output to the web UI SSE stream.
pub type OutputConsumer = Box<dyn Fn(&str) + Send + Sync>;

/// Processes Claude Code's stream-json output.
/// Parses JSON events, logs key information with tracing, and forwards
/// formatted output to the consumer for real-time web UI display.
pub struct ClaudeMessageProcessor {
    consumer: Option<OutputConsumer>,
}

impl ClaudeMessageProcessor {
    /// Create a new processor with an optional output consumer.
    pub fn new(consumer: Option<OutputConsumer>) -> Self {
        Self { consumer }
    }

    /// Process a single line of output from Claude Code.
    pub fn process(&self, line: &str) {
        // Try to parse as JSON event
        if let Ok(json) = serde_json::from_str::<Value>(line) {
            if let Some(msg_type) = json.get("type").and_then(|v| v.as_str()) {
                match msg_type {
                    "stream_event" => self.handle_stream_event(&json),
                    "assistant" => self.handle_assistant(&json),
                    "user" => self.handle_user_message(&json),
                    "system" => self.handle_system(&json),
                    "result" => self.handle_result(&json),
                    _ => self.fallback(line),
                }
            } else {
                self.fallback(line);
            }
        } else {
            // Not JSON - likely a warning or raw output from Claude
            self.fallback(line);
        }
    }

    fn handle_stream_event(&self, json: &Value) {
        if let Some(event) = json.get("event").and_then(|v| v.as_object()) {
            if let Some(event_type) = event.get("type").and_then(|v| v.as_str()) {
                match event_type {
                    "message_start" => self.handle_message_start(event),
                    "message_delta" => {
                        // message_delta with stop_reason - not much to log
                    }
                    "content_block_delta" => self.handle_content_block_delta(event),
                    "content_block_start" => {} // ignore
                    "content_block_stop" => {}  // ignore
                    "message_stop" => self.send_output("\n"),
                    _ => self.send_output(&json.to_string()),
                }
            } else {
                self.send_output(&json.to_string());
            }
        } else {
            self.send_output(&json.to_string());
        }
    }

    fn handle_message_start(&self, event: &serde_json::Map<String, Value>) {
        if let Some(message) = event.get("message").and_then(|v| v.as_object()) {
            if let Some(content) = message.get("content") {
                if content.is_array() {
                    if let Some(items) = content.as_array() {
                        for item in items {
                            if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                match item_type {
                                    "thinking" => {
                                        if let Some(thinking) = item.get("thinking").and_then(|v| v.as_str()) {
                                            self.send_output(thinking);
                                        }
                                    }
                                    "text" => {
                                        if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                            self.send_output(text);
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }
                    }
                } else if let Some(content_type) = content.get("type").and_then(|v| v.as_str()) {
                    match content_type {
                        "thinking" => {
                            if let Some(thinking) = content.get("thinking").and_then(|v| v.as_str()) {
                                self.send_output(thinking);
                            }
                        }
                        "text" => {
                            if let Some(text) = content.get("text").and_then(|v| v.as_str()) {
                                self.send_output(text);
                            }
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    fn handle_content_block_delta(&self, event: &serde_json::Map<String, Value>) {
        if let Some(delta) = event.get("delta").and_then(|v| v.as_object()) {
            if let Some(delta_type) = delta.get("type").and_then(|v| v.as_str()) {
                match delta_type {
                    "thinking_delta" => {
                        if let Some(thinking) = delta.get("thinking").and_then(|v| v.as_str()) {
                            self.send_output(thinking);
                        }
                    }
                    "text_delta" => {
                        if let Some(text) = delta.get("text").and_then(|v| v.as_str()) {
                            self.send_output(text);
                        }
                    }
                    "input_json_delta" => {
                        // Partial JSON for tool calls - skip for cleaner output
                    }
                    "signature_delta" => {
                        // Skip signature deltas
                    }
                    _ => {}
                }
            }
        }
    }

    fn handle_assistant(&self, json: &Value) {
        if let Some(message) = json.get("message").and_then(|v| v.as_object()) {
            if let Some(content) = message.get("content") {
                if content.is_array() {
                    if let Some(items) = content.as_array() {
                        for item in items {
                            if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                match item_type {
                                    "thinking" => {
                                        if let Some(thinking) = item.get("thinking").and_then(|v| v.as_str()) {
                                            self.send_output(thinking);
                                        }
                                    }
                                    "tool_use" => self.render_tool_use(item),
                                    "text" => {
                                        if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                            self.send_output(text);
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }
                    }
                } else if let Some(content_type) = content.get("type").and_then(|v| v.as_str()) {
                    match content_type {
                        "thinking" => {
                            if let Some(thinking) = content.get("thinking").and_then(|v| v.as_str()) {
                                self.send_output(thinking);
                            }
                        }
                        "tool_use" => self.render_tool_use(content),
                        "text" => {
                            if let Some(text) = content.get("text").and_then(|v| v.as_str()) {
                                self.send_output(text);
                            }
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    fn handle_user_message(&self, json: &Value) {
        if let Some(message) = json.get("message").and_then(|v| v.as_object()) {
            if let Some(content) = message.get("content") {
                if content.is_array() {
                    if let Some(items) = content.as_array() {
                        for item in items {
                            if let Some(item_type) = item.get("type").and_then(|v| v.as_str()) {
                                match item_type {
                                    "text" => {
                                        if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                                            self.send_output(&format!("[User]: {}\n", text));
                                        }
                                    }
                                    "tool_result" => {
                                        if let Some(tool_content) = item.get("content").and_then(|v| v.as_str()) {
                                            let with_newlines = tool_content.replace("\\n", "\n");
                                            info!("[Tool result] {}", crate::safe_truncate(&with_newlines, 500));
                                            self.send_output(&format!("tool_result:\n{}\n", with_newlines));
                                        } else {
                                            self.send_output(&format!("tool_result: {}\n", item));
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }
                    }
                } else if let Some(content_type) = content.get("type").and_then(|v| v.as_str()) {
                    match content_type {
                        "text" => {
                            if let Some(text) = content.get("text").and_then(|v| v.as_str()) {
                                self.send_output(&format!("[User]: {}\n", text));
                            }
                        }
                        "tool_result" => {
                            if let Some(tool_content) = content.get("content").and_then(|v| v.as_str()) {
                                let with_newlines = tool_content.replace("\\n", "\n");
                                info!("[Tool result] {}", crate::safe_truncate(&with_newlines, 500));
                                self.send_output(&format!("tool_result:\n{}\n", with_newlines));
                            } else {
                                self.send_output(&format!("tool_result: {}\n", content));
                            }
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    fn handle_system(&self, _json: &Value) {
        // System messages - log at debug level
        debug!("Received system message from Claude Code");
    }

    fn handle_result(&self, json: &Value) {
        // Result event contains session summary: cost, tokens, turns, etc.
        let cost = json.get("total_cost_usd").and_then(|v| v.as_f64()).unwrap_or(0.0);
        let turns = json.get("num_turns").and_then(|v| v.as_i64()).unwrap_or(0);
        info!("[Claude Code result] Total cost: ${:.4}, Total turns: {}", cost, turns);
        self.send_output(&format!("[Result] cost=${:.4}, turns={}\n", cost, turns));
    }

    fn render_tool_use(&self, item: &Value) {
        if let Some(name) = item.get("name").and_then(|v| v.as_str()) {
            match name {
                "Edit" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        let file_path = input.get("file_path").and_then(|v| v.as_str()).unwrap_or("");
                        let old_string = input.get("old_string").and_then(|v| v.as_str()).unwrap_or("");
                        let new_string = input.get("new_string").and_then(|v| v.as_str()).unwrap_or("");
                        info!("[Tool: Edit] {}", file_path);
                        let old_normalized = old_string.replace("\\n", "\n");
                        let new_normalized = new_string.replace("\\n", "\n");
                        self.send_output(&format!("Edit {}\nOld\n{}\nNew\n{}\n", file_path, old_normalized, new_normalized));
                    }
                }
                "Glob" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        if let Some(pattern) = input.get("pattern").and_then(|v| v.as_str()) {
                            info!("[Tool: Glob] pattern={}", pattern);
                            self.send_output(&format!("\ntool_use: Glob {}\n", pattern));
                        }
                    }
                }
                "Grep" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        let pattern = input.get("pattern").and_then(|v| v.as_str()).unwrap_or("");
                        let file_path = input.get("file_path").and_then(|v| v.as_str()).unwrap_or("");
                        info!("[Tool: Grep] pattern={} file={}", pattern, file_path);
                        self.send_output(&format!("\ntool_use: Grep {} in {}\n", pattern, file_path));
                    }
                }
                "Read" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        if let Some(file_path) = input.get("file_path").and_then(|v| v.as_str()) {
                            info!("[Tool: Read] {}", file_path);
                            self.send_output(&format!("\ntool_use: Read {}\n", file_path));
                        }
                    }
                }
                "Write" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        let file_path = input.get("file_path").and_then(|v| v.as_str()).unwrap_or("");
                        let content = input.get("content").and_then(|v| v.as_str()).unwrap_or("");
                        info!("[Tool: Write] {}", file_path);
                        let normalized = content.replace("\\n", "\n");
                        self.send_output(&format!("\ntool_use: Write {}\nContent:\n{}\n", file_path, normalized));
                    }
                }
                "Bash" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        let run_in_background = input.get("run_in_background").and_then(|v| v.as_bool()).unwrap_or(false);
                        let description = input.get("description").and_then(|v| v.as_str()).unwrap_or("");
                        let command = input.get("command").and_then(|v| v.as_str()).unwrap_or("");
                        let bg = if run_in_background { " (in background)" } else { "" };
                        info!("[Tool: Bash{}] {} - {}", bg, description, command);
                        self.send_output(&format!("\ntool_use: Bash{} {}\nCommand: {}\n", bg, description, command));
                    }
                }
                "Task" => {
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        let prompt = input.get("prompt").and_then(|v| v.as_str()).unwrap_or("");
                        info!("[Tool: Task] {}", crate::safe_truncate(&prompt, 200));
                        self.send_output(&format!("\ntool_use: Task {}\n", crate::safe_truncate(&prompt, 200)));
                    }
                }
                "TodoWrite" => {
                    info!("[Tool: TodoWrite]");
                    self.send_output("\ntool_use: TodoWrite\n");
                    if let Some(input) = item.get("input").and_then(|v| v.as_object()) {
                        if let Some(todos) = input.get("todos").and_then(|v| v.as_array()) {
                            for todo in todos {
                                let content = todo.get("content").and_then(|v| v.as_str()).unwrap_or("");
                                let status = todo.get("status").and_then(|v| v.as_str()).unwrap_or("");
                                match status {
                                    "in_progress" => self.send_output(&format!("[⟳] {}\n", content)),
                                    "pending" => self.send_output(&format!("[⌛] {}\n", content)),
                                    "completed" => self.send_output(&format!("[✅] {}\n", content)),
                                    _ => self.send_output(&format!("[ ] {}\n", content)),
                                }
                            }
                        }
                    }
                }
                _ => {
                    debug!("[Tool: {}] args={}", name, item.get("input").map(|i| i.to_string()).unwrap_or_default());
                    self.send_output(&format!("\ntool_use: {}\n", name));
                }
            }
        }
    }

    fn fallback(&self, line: &str) {
        // Not a recognized JSON event - print as-is (warnings, errors, etc.)
        info!("[Claude output] {}", line);
        self.send_output(line);
    }

    fn send_output(&self, text: &str) {
        if let Some(ref consumer) = self.consumer {
            consumer(text);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Regression: tool results / prompts containing multi-byte UTF-8 used to
    /// panic the byte-slice truncation in the log path. With a subscriber at
    /// INFO the format args are evaluated, so this must not panic.
    fn with_info_subscriber(f: impl FnOnce()) {
        let _guard = tracing::subscriber::set_default(
            tracing_subscriber::fmt()
                .with_max_level(tracing::Level::INFO)
                .with_writer(std::io::sink)
                .finish(),
        );
        f();
    }

    #[test]
    fn non_ascii_tool_result_does_not_panic() {
        with_info_subscriber(|| {
            let processor = ClaudeMessageProcessor::new(None);
            // >500 bytes, multi-byte chars straddling the 500-byte limit
            let json = format!(
                r#"{{"type":"user","message":{{"content":[{{"type":"tool_result","tool_use_id":"t1","content":"{}"}}]}}}}"#,
                "€".repeat(167)
            );
            processor.process(&json);
        });
    }

    #[test]
    fn non_ascii_task_prompt_does_not_panic() {
        with_info_subscriber(|| {
            let processor = ClaudeMessageProcessor::new(None);
            // >200 bytes, multi-byte chars straddling the 200-byte limit
            let json = format!(
                r#"{{"type":"assistant","message":{{"content":[{{"type":"tool_use","name":"Task","input":{{"prompt":"{}"}}}}]}}}}"#,
                "€".repeat(67)
            );
            processor.process(&json);
        });
    }
}
